### PR TITLE
[Whisper] Fix audio classification with weighted layer sum

### LINKED ISSUE #28002: Not handled case when use_weighted_layer_sum and return-dict=True in WhisperForAudioClassification
@sanchit-gandhi 
I use WhisperForAudioClassification task and want to use `use_weighted_layer_sum=True`, but there is a problem when call forward, 
the encoder part can return tuple or dict if `return_dict=True` but the code for use `use_weighted_layer_sum=True` assume the return to be tuple only and this line raise error `hidden_states = torch.stack(encoder_outputs, dim=1)` if the encoder return dict, there are workaround by using `return_dict=False` but when use the model later with `pipeline` it will raise error because it assume the model to return dict not tuple. [Link to code with the problem](https://github.com/huggingface/transformers/blob/c7f076a00ee54f777b3d3322c91bc11489a47950/src/transformers/models/whisper/modeling_whisper.py#L2918C6-L2918C6)
```py
        if self.config.use_weighted_layer_sum:
            hidden_states = torch.stack(encoder_outputs, dim=1) # This line raise error when return_dict=True and use_weighted_layer_sum=True
            norm_weights = nn.functional.softmax(self.layer_weights, dim=-1)
            hidden_states = (hidden_states * norm_weights.view(-1, 1, 1)).sum(dim=1)
        else:
            hidden_states = encoder_outputs[0]
```
**Reproduce error**
```py
from transformers import  WhisperForAudioClassification, AutoFeatureExtractor
from datasets import load_dataset

dataset = load_dataset('seba3y/speechocean762',)
dataset = dataset['train']
sampling_rate = dataset.features["audio"].sampling_rate
dataset = dataset.remove_columns(['utt_name', 'text', 'completeness', 'fluency', 'prosodic'])


feature_extractor = AutoFeatureExtractor.from_pretrained("seba3y/whisper-tiny")
model = WhisperForAudioClassification.from_pretrained("seba3y/whisper-tiny",
                                                       use_weighted_layer_sum=True, 
                                                       return_dict=True)
# test if it work
inputs = feature_extractor(dataset['train'][3]["audio"]["array"], sampling_rate=sampling_rate, return_tensors="pt")
with torch.no_grad():
    logits = model(**inputs).logits
predicted_class_ids = torch.argmax(logits, dim=-1).item()
predicted_label = model.config.id2label[predicted_class_ids]
print(predicted_label)
```

### PR DESCRIPTION
**Pull request by sanchit-gandhi**
# What does this PR do?

Fixes #28002: the `WhisperForAudioClassfication` is corrected for compatibility with `use_weighted_layer_sum=True`. Implements 2 tests to confirm correctness.

### CODE REVIEW

File: src/transformers/models/whisper/modeling_whisper.py [hunk 1]
```diff
@@ -2957,6 +2959,11 @@ def forward(
         output_hidden_states = (
             output_hidden_states if output_hidden_states is not None else self.config.output_hidden_states
         )
+        if self.config.use_weighted_layer_sum:
```
Comment (sanchit-gandhi) [inline src/transformers/models/whisper/modeling_whisper.py:2962]: Following the convention we use in Wav2Vec2 by forcing `output_hidden_states=True` when `use_weighted_layer_sum=True`: 
https://github.com/huggingface/transformers/blob/2c1eebc1216549d8195d7d1c6adb8b99afee3ec5/src/transformers/models/wav2vec2/modeling_wav2vec2.py#L2093

File: src/transformers/models/whisper/modeling_whisper.py [hunk 2]
```diff
@@ -2969,7 +2976,8 @@ def forward(
             )
 
         if self.config.use_weighted_layer_sum:
-            hidden_states = torch.stack(encoder_outputs, dim=1)
+            hidden_states = encoder_outputs[_HIDDEN_STATES_START_POSITION]
```
Comment (sanchit-gandhi) [inline src/transformers/models/whisper/modeling_whisper.py:2979]: Again, following the Wav2Vec2 design:
https://github.com/huggingface/transformers/blob/2c1eebc1216549d8195d7d1c6adb8b99afee3ec5/src/transformers/models/wav2vec2/modeling_wav2vec2.py#L2104

**Review by amyeroberts — APPROVED** [review]
Thanks for fixing!
